Introduce public MergedDataMap

lucylq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #14766

Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

lucylq committed:
Introduce public MergedDataMap

Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

swolchok commented on extension/named_data_map/merged_data_map.cpp:
```diff
@@ -0,0 +1,117 @@
+using executorch::runtime::FreeableBuffer;
+using executorch::runtime::Result;
+
+namespace executorch {
+namespace extension {
+
```
nit: `namespace executorch::extension {`

swolchok reviewed (changes requested):
ping me when this is ready for re-review so you don't have to wait

lucylq commented on extension/named_data_map/merged_data_map.h:
```diff
@@ -0,0 +1,107 @@
+   * Note: the data maps must outlive the MergedDataMap instance.
+   */
+  static executorch::runtime::Result<MergedDataMap> load(
+      std::vector<const executorch::ET_RUNTIME_NAMESPACE::NamedDataMap*>
```
Thanks - will update to use a span

lucylq commented on extension/named_data_map/merged_data_map.cpp:
```diff
@@ -0,0 +1,117 @@
+    uint32_t num_keys = cur_map->get_num_keys().get();
+    for (uint32_t j = 0; j < num_keys; ++j) {
+      const auto cur_key = cur_map->get_key(j).get();
+      ET_CHECK_OR_RETURN_ERROR(
+          key_to_map_index.find(cur_key) == key_to_map_index.end(),
+          InvalidArgument,
+          "Duplicate key %s in named data maps at index %u and %u",
+          cur_key,
+          key_to_map_index.at(cur_key),
+          i);
+      key_to_map_index[cur_key] = i;
```
Thanks, this is great!

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

swolchok commented on extension/named_data_map/test/CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+include(${EXECUTORCH_ROOT}/tools/cmake/Test.cmake)
+
+add_custom_command(
+  OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/ModuleAddMulProgram.pte"
+         "${CMAKE_CURRENT_BINARY_DIR}/ModuleAddMulProgram.ptd"
```
nit: mildly surprised cmake-format doesn't alphabetize these

swolchok commented on extension/named_data_map/test/merged_data_map_test.cpp:
```diff
@@ -0,0 +1,197 @@
+  void load_flat_tensor_data_map(const char* path, const char* module_name) {
+    Result<FileDataLoader> loader = FileDataLoader::from(path);
+    ASSERT_EQ(loader.error(), Error::Ok);
+    loaders_.insert(
```
nit: emplace removes the need for the braces

swolchok commented on extension/named_data_map/test/merged_data_map_test.cpp:
```diff
@@ -0,0 +1,197 @@
+        FlatTensorDataMap::load(loaders_[module_name].get());
+    EXPECT_EQ(data_map.error(), Error::Ok);
+
+    data_maps_.insert(
```
ditto

swolchok commented on extension/named_data_map/test/merged_data_map_test.cpp:
```diff
@@ -0,0 +1,197 @@
+  EXPECT_EQ(layout1.scalar_type(), layout2.scalar_type());
+  EXPECT_EQ(layout1.nbytes(), layout2.nbytes());
+  EXPECT_EQ(layout1.sizes().size(), layout2.sizes().size());
+  for (size_t i = 0; i < layout1.sizes().size(); i++) {
```
super nit: irange

swolchok commented on extension/named_data_map/test/merged_data_map_test.cpp:
```diff
@@ -0,0 +1,197 @@
+
+    // Compare load_into.
+    auto nbytes = ndm_meta.nbytes();
+    void* ndm_buffer = malloc(nbytes);
```
use unique_ptr (with an array if needed) instead of raw malloc whenever possible

swolchok commented on extension/named_data_map/test/merged_data_map_test.cpp:
```diff
@@ -0,0 +1,197 @@
+      data_maps_["addmul"].get(), data_maps_["simple_train"].get()};
+  Result<MergedDataMap> merged_map2 = MergedDataMap::load(
+      Span<const NamedDataMap*>(ndms2.data(), ndms2.size()));
+  EXPECT_EQ(merged_map2.error(), Error::Ok);
```
what about examining the contents of the merged map to make sure the merge is happening correctly? I think that's partially happening in CheckDataMapContents below; why not do it here?

lucylq replied:
thanks - combined the test cases

swolchok commented on extension/named_data_map/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Please this file formatted by running:
```
this sentence no verb

swolchok commented on extension/named_data_map/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+install(
+  TARGETS extension_named_data_map
+  EXPORT ExecuTorchTargets
+  DESTINATION lib
```
this should be CMAKE_INSTALL_LIBDIR from GNUInstallDirs; inconveniently I never landed #12792 so we have a bunch of bad examples laying around

swolchok commented on extension/named_data_map/merged_data_map.cpp:
```diff
@@ -0,0 +1,117 @@
+
+  // Check for duplicate keys.
+  std::unordered_map<std::string, uint32_t> key_to_map_index;
+  for (uint32_t i = 0; i < valid_data_maps.size(); i++) {
```
irange

swolchok commented on extension/named_data_map/merged_data_map.cpp:
```diff
@@ -0,0 +1,117 @@
+  for (uint32_t i = 0; i < valid_data_maps.size(); i++) {
+    const auto cur_map = valid_data_maps[i];
+    uint32_t num_keys = cur_map->get_num_keys().get();
+    for (uint32_t j = 0; j < num_keys; ++j) {
```
irange

swolchok reviewed (approved):
just a bunch of minor complaints; seems fine overall

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]

lucylq committed:
Update on "Introduce public MergedDataMap"


Add public merged data map. Module can use this to resolve multiple named data maps.

Differential Revision: [D83527299](https://our.internmc.facebook.com/intern/diff/D83527299/)

[ghstack-poisoned]
